tests/cephfs/snapshot_clone/test_referent_inode_singleFS.py

```diff
@@ -0,0 +1,458 @@
+):
+    """
+    Creates a hardlink, fetches file stats, validates inode numbers,
+    and checks for the newly created RADOS object.
+    """
+
+    def exec_stat(file_path):
+        """Executes stat command and extracts inode number."""
+        stdout, stderr = client.exec_command(sudo=True, cmd=f"stat {file_path}")
+        log.info(f"File statistics for {file_path}:\n{stdout}")
+        match = re.search(r"Inode:\s+(\d+)", stdout)
+        return match.group(1) if match else None
+
+    # Step 1: Capture initial RADOS objects
+    initial_rados_objects = set(get_rados_object_from_datapool(client, data_pool_name))
+    log.info(f"Initial RADOS objects: {initial_rados_objects}")
+
+    # Step 2: Create hard link
+    client.exec_command(sudo=True, cmd=f"ln {file1_path} {hl_file_path}")
+    log.info(f"Created hardlink: {hl_file_path}")
+
+    # Step 3: Flush journal to ensure changes are recorded
+    flush_journal_on_active_mdss(fs_util, client, fs_name)
+
+    # Step 4: Compare Inode to validate hardlinks
+    inode1, inode2 = exec_stat(file1_path), exec_stat(hl_file_path)
+    if inode1 and inode2 and inode1 == inode2:
+        log.info(f"Validation Passed: Inodes match ({inode1})")
+    else:
+        log.error(f"Validation Failed: Inodes do not match ({inode1} != {inode2})")
+        return False, "no", None
+
+    # Step 5: Capture RADOS objects after creating hardlink
+    post_rados_objects = set(get_rados_object_from_datapool(client, data_pool_name))
+    log.info(f"RADOS objects after creating hardlink: {post_rados_objects}")
+
+    # Step 6: Identify the newly created RADOS object
+    new_rados_objects = post_rados_objects - initial_rados_objects
+    log.info(f"Newly created RADOS objects: {new_rados_objects}")
+
+    # Step 7: Validate referent inode
+    referent_inode = "yes" if len(new_rados_objects) > 0 else "no"
+    log.info(f"Referent Inode: {referent_inode}")
+
+    return True, referent_inode, new_rados_objects
+
+
+def flush_journal_on_active_mdss(fs_util, client, fs_name):
+    # Call get_active_mdss within this function to get the list of active MDS nodes
+    active_mdss = fs_util.get_active_mdss(client, fs_name)
+    for mds in active_mdss:
+        client.exec_command(sudo=True, cmd=f"ceph tell mds.{mds} flush journal")
+        log.info(f"Journal flushed on MDS: mds.{mds}")
+
+
+def validate_referent_inode(
+    fs_util,
+    client,
+    mount_path,
+    data_pool,
+    fs_name,
+):
+    """
+    Automate the snapshot creation and testing.
+    """
+    log.info("Starting snapshot automation...")
+
+    # Step 1: Create directories
+    dirs = ["dir1", "dir2", "dir3"]
+    create_directories(client, mount_path, dirs)
+    flush_journal_on_active_mdss(fs_util, client, fs_name)
+
+    # Step 2: Create files and snapshots in dir1
+    file1_path = f"{mount_path}/dir1/file1"
+    create_file_with_content(client, file1_path, "first file")
+    flush_journal_on_active_mdss(fs_util, client, fs_name)
```

Comment:
Do we need to flush journal on every operation? What is the default time at which journal flush is handled automatically?

Reply:
Since we are verifying the RADOS objects immediately after file creation, flushing is necessary to ensure they are captured in automation. Occasionally, the objects were not detected after snapshot creation and linking, leading to failures in capturing them in automation